rust/src/tests/chflags.rs

```diff
@@ -0,0 +1,218 @@
+use std::{collections::HashSet, iter::once};
+
+use nix::{
+    errno::Errno,
+    libc::fflags_t,
+    sys::stat::{lstat, stat, FileFlag},
+    unistd::chflags,
+};
+use once_cell::sync::Lazy;
+
+#[cfg(any(target_os = "netbsd", target_os = "freebsd", target_os = "dragonfly"))]
+use crate::utils::lchflags;
+use crate::{
+    runner::context::{FileType, SerializedTestContext, TestContext},
+    test::{FileFlags, FileSystemFeature},
+};
+
+use super::{assert_ctime_changed, assert_ctime_unchanged};
+
+//TODO: Does the split user/system flags make sense? Besides, only FreeBSD ones are specified in the original test suite.
+
+const USER_FLAGS: Lazy<HashSet<FileFlags>> = Lazy::new(|| {
+    HashSet::from([
+        FileFlags::UF_NODUMP,
+        FileFlags::UF_IMMUTABLE,
+        FileFlags::UF_APPEND,
+        #[cfg(any(target_os = "freebsd", target_os = "dragonfly"))]
+        FileFlags::UF_NOUNLINK,
+        #[cfg(any(target_os = "freebsd", target_os = "dragonfly"))]
+        FileFlags::UF_OPAQUE,
+    ])
+});
+
+const SYSTEM_FLAGS: Lazy<HashSet<FileFlags>> = Lazy::new(|| {
+    HashSet::from([
+        FileFlags::SF_ARCHIVED,
+        FileFlags::SF_IMMUTABLE,
+        FileFlags::SF_APPEND,
+        #[cfg(any(target_os = "freebsd", target_os = "dragonfly"))]
+        FileFlags::SF_NOUNLINK,
+    ])
+});
+
+fn get_flags(ctx: &TestContext) -> (FileFlag, FileFlag, FileFlag) {
+    let allflags: FileFlag = ctx
+        .features_config()
+        .file_flags
+        .iter()
+        .copied()
+        .map(Into::into)
+        .collect();
+
+    let user_flags: FileFlag = ctx
+        .features_config()
+        .file_flags
+        .intersection(&USER_FLAGS)
+        .copied()
+        .map(Into::into)
+        .collect();
+
+    let system_flags: FileFlag = ctx
+        .features_config()
+        .file_flags
+        .intersection(&SYSTEM_FLAGS)
+        .copied()
+        .map(Into::into)
+        .collect();
+
+    (allflags, user_flags, system_flags)
+}
+
+crate::test_case! {
+    /// chflags(2) set the flags provided for the file.
+    set_flags, FileSystemFeature::Chflags => [Regular, Dir, Fifo, Block, Char, Socket]
+}
+fn set_flags(ctx: &mut TestContext, ft: FileType) {
+    let (flags, user_flags, system_flags) = get_flags(ctx);
+
+    let file = ctx.create(ft.clone()).unwrap();
+    let file_flags = stat(&file).unwrap().st_flags;
+    assert_eq!(file_flags, FileFlag::empty().bits() as fflags_t);
+
+    for flags_set in [flags, user_flags, system_flags, FileFlag::empty()] {
+        assert!(chflags(&file, flags_set).is_ok());
+        let file_flags = stat(&file).unwrap().st_flags;
+        assert_eq!(file_flags, flags_set.bits() as fflags_t);
+    }
+
+    // Check with lchflags
+
+    let file = ctx.create(ft).unwrap();
```

Comment:
This part is pretty useless, because symlink resolution happens way above the FS layer.  But it _was_ in the original, so I guess we should keep it.